src/ucp/core/ucp_worker.c

```diff
@@ -2350,3 +2356,90 @@ void ucp_worker_print_info(ucp_worker_h worker, FILE *stream)
 
     UCP_WORKER_THREAD_CS_EXIT_CONDITIONAL(worker);
 }
+
+static int ucp_worker_keepalive_is_enabled(ucp_worker_h worker)
+{
+    return (worker->context->config.ext.keepalive_timeout != 0.0) &&
+           (worker->context->config.ext.keepalive_num_eps != 0);
+}
+
+static unsigned ucp_worker_keepalive_cb(void *arg)
+{
+    ucp_worker_h worker = (ucp_worker_h)arg;
+    double now          = ucs_time_to_sec(ucs_get_time());
+    unsigned cnt;
+    ucp_ep_ext_gen_t *ep_ext;
+    ucs_list_link_t *iter_begin;
+
+    if (ucs_likely((now - worker->keepalive.last_shot) <
+                   worker->context->config.ext.keepalive_timeout)) {
+        return 0;
+    }
+
+    worker->keepalive.last_shot = now;
+    cnt                         = 0;
+    iter_begin                  = worker->keepalive.iter;
+    /* use own loop for elements because standard for_each skips
+     * head element */
+    do {
+        ep_ext = ucs_list_next(worker->keepalive.iter,
```

Comment:
please add TODO somewhere that separate list is needed. This list would contain only those eps, which are configured with error handling (to avoid traversing thru all the eps)

Reply:
added TODO note